compiler_opt/distributed/local/buffered_scheduler.py

```diff
@@ -0,0 +1,82 @@
+# coding=utf-8
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""An optimal push-pull-based load balancer which attempts to maintain at least
+`buffer` tasks assigned to each assignee.
+"""
+
+import concurrent.futures
+import threading
+
+from typing import List, Callable, TypeVar
+
+
+class DeadlockDetected(Exception):
+
+  def __init__(self):
+    Exception.__init__(self)
+
+
+def raise_deadlock(*args, **kwargs):
+  # This is raised if attempting to call .result() on _NEVER_FUTURE
+  raise DeadlockDetected()
+
+
+_NEVER_FUTURE = concurrent.futures.Future()  # A Future that never completes.
+_NEVER_FUTURE.result = raise_deadlock
+T = TypeVar('T')
+
+
+def schedule(assignments: List[Callable[[T], concurrent.futures.Future]],
+             assignees: List[T],
+             buffer=2) -> List[concurrent.futures.Future]:
+  """
+  Assigns assignments to assignees once previous assignments of the assignee are
+  completed.
+  Args:
+    assignments: Function to call with an assignee.
+    assignees: List of assignees that are the singular argument to callable.
+    buffer: Number of assignments to maintain on each assignee.
+  Returns:
+    A list of Futures that should be stored AS IS. Treat as a reference-to-list.
+  """
+  # Create "fake" futures to be returned first, these futures will be replaced
+  # with "real" futures later on.
+  results = [_NEVER_FUTURE] * len(assignments)
+  idx = -1
+  idx_lock = threading.Lock()
+
+  # Simple atomic increment and get.
+  def fetch_idx():
```

Comment:
why not just use a queue that gets drained until empty()?

If the reason is "why allocate more", sgtm, and I'd then comment that `fetch_idx` + the list of work form a thread_safely dequeable queue, quicker to understand (i.e. the fact that fetch_idx does simple atomic increment is quite immediately obvious, but "why" is less so until one finishes reading the code)